File: sysstat/sanitizer.go
package sysstat

import (
	"strings"
)

var sensitiveFlags = map[string]bool{
	"--password":      true,
	"-p":              false, // -p is ambiguous (e.g., port), so we don't redact it automatically.
	"--token":         true,
	"--api-key":       true,
	"--apikey":        true,
	"--secret":        true,
	"--client-secret": true,
	"--access-token":  true,
	"--auth-token":    true,
	"--session-id":    true,
}

var sensitiveSuffixes = []string{
	"password",
	"token",
	"secret",
	"_key", // Matches api_key, access_key, but not keyboard
	"-key", // Matches ssh-key, private-key
	".key", // Matches file.key
	"signature",
	"credential",
	"cookie",
	"session",
}

// SanitizeCommand joins command arguments into a string, redacting sensitive information.
func SanitizeCommand(args []string) string {
	if len(args) == 0 {
		return ""
	}

	var sanitized []string
	skipNext := false

	for i, arg := range args {
		if skipNext {
			sanitized = append(sanitized, "[REDACTED]")
			skipNext = false
			continue
		}

		// Check for --flag=value
		if strings.Contains(arg, "=") {
			parts := strings.SplitN(arg, "=", 2)
			key := parts[0]
			// val := parts[1] // unused

			// Normalize key for checking (remove leading dashes)
			cleanKey := strings.TrimLeft(key, "-")

			if isSensitiveKey(cleanKey) {
				sanitized = append(sanitized, key+"=[REDACTED]")
				continue
			}

			// Also check if the key matches a sensitive flag explicitly
			// We must check lowercased key to handle case variations
			if sensitiveFlags[strings.ToLower(key)] {
				sanitized = append(sanitized, key+"=[REDACTED]")
				continue
			}

			// If not sensitive, keep as is
			sanitized = append(sanitized, arg)
			continue
		}

		// Check for sensitive flags that take the next argument
		// 1. Check strict list (case-insensitive)
		lowerArg := strings.ToLower(arg)
		if val, ok := sensitiveFlags[lowerArg]; ok && val {
			sanitized = append(sanitized, arg)
			if i+1 < len(args) {
				skipNext = true
			}
			continue
		}

		// 2. Check heuristics (suffix matching)
		// Clean the arg (remove leading dashes)
		cleanArg := strings.TrimLeft(arg, "-")
		if isSensitiveKey(cleanArg) {
			sanitized = append(sanitized, arg)
			// Only redact next if it doesn't look like another flag
			// This prevents false positives for boolean flags (e.g., --enable-password-auth --verbose)
			if i+1 < len(args) && !strings.HasPrefix(args[i+1], "-") {
				skipNext = true
			}
			continue
		}

		sanitized = append(sanitized, arg)
	}

	return strings.Join(sanitized, " ")
}

func isSensitiveKey(key string) bool {
	lowerKey := strings.ToLower(key)

	// Exact matches
	if lowerKey == "password" || lowerKey == "token" || lowerKey == "secret" || lowerKey == "key" || lowerKey == "auth" {
		return true
	}

	// Suffix matches
	for _, suffix := range sensitiveSuffixes {
		if strings.HasSuffix(lowerKey, suffix) {
			// If the match is the entire string, it's a match
			if len(lowerKey) == len(suffix) {
				return true
			}

			// If the suffix itself starts with a separator, it implies a boundary
			if suffix[0] == '-' || suffix[0] == '_' || suffix[0] == '.' {
				return true
			}

			// Otherwise, check if the suffix is preceded by a separator
			matchIndex := len(lowerKey) - len(suffix)
			if matchIndex > 0 {
				charBefore := lowerKey[matchIndex-1]
				if charBefore == '-' || charBefore == '_' || charBefore == '.' {
					return true
				}
			}
		}
	}
	return false
}

File: rust/src/sysstat/sanitizer.rs
use once_cell::sync::Lazy;
use std::collections::HashMap;

static SENSITIVE_FLAGS: Lazy<HashMap<&'static str, bool>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("--password", true);
    m.insert("-p", false); // Ambiguous
    m.insert("--token", true);
    m.insert("--api-key", true);
    m.insert("--apikey", true);
    m.insert("--secret", true);
    m.insert("--client-secret", true);
    m.insert("--access-token", true);
    m.insert("--auth-token", true);
    m
});

static SENSITIVE_SUFFIXES: Lazy<Vec<&'static str>> =
    Lazy::new(|| vec![
        "password",
        "token",
        "secret",
        "_key",
        "-key",
        ".key",
        "signature",
        "credential",
        "cookie",
        "session",
    ]);

// sanitize_command reconstructs the command line string from arguments while redacting sensitive information.
// It aims for parity with the Go implementation, handling both `--flag=value` and `--flag value` patterns.
// Note: Space-separated flags rely on a hardcoded allowlist (SENSITIVE_FLAGS), while `=` separated flags
// use heuristic matching for keys (e.g. ending in "password" or "token").
pub fn sanitize_command(args: &[String]) -> String {
    if args.is_empty() {
        return String::new();
    }

    let mut sanitized = Vec::new();
    let mut skip_next = false;

    for (i, arg) in args.iter().enumerate() {
        if skip_next {
            sanitized.push("[REDACTED]".to_string());
            skip_next = false;
            continue;
        }

        // Check for --flag=value
        if let Some((key, _)) = arg.split_once('=') {
            // Normalize key (remove leading dashes)
            let clean_key = key.trim_start_matches('-');

            if is_sensitive_key(clean_key) {
                sanitized.push(format!("{}=[REDACTED]", key));
                continue;
            }

            // Check if key matches a sensitive flag explicitly
            if SENSITIVE_FLAGS.contains_key(key.to_lowercase().as_str()) {
                sanitized.push(format!("{}=[REDACTED]", key));
                continue;
            }

            sanitized.push(arg.clone());
            continue;
        }

        let lower_arg = arg.to_lowercase();
        // Check for sensitive flags that take the next argument
        if let Some(&should_skip) = SENSITIVE_FLAGS.get(lower_arg.as_str()) {
            sanitized.push(arg.clone());
            if should_skip && i + 1 < args.len() {
                skip_next = true;
            }
            continue;
        }

        let clean_arg = lower_arg.trim_start_matches('-');
        if is_sensitive_key(clean_arg) {
            sanitized.push(arg.clone());
            if i + 1 < args.len() && !args[i + 1].starts_with('-') {
                skip_next = true;
            }
            continue;
        }

        sanitized.push(arg.clone());
    }

    sanitized.join(" ")
}

fn is_sensitive_key(key: &str) -> bool {
    let lower_key = key.to_lowercase();

    // Exact matches
    if matches!(
        lower_key.as_str(),
        "password" | "token" | "secret" | "key" | "auth"
    ) {
        return true;
    }

    // Suffix matches
    for suffix in SENSITIVE_SUFFIXES.iter() {
        if lower_key.ends_with(suffix) {
            // If the match is the entire string, it's a match
            if lower_key.len() == suffix.len() {
                return true;
            }

            // If the suffix itself starts with a separator, it implies a boundary
            if suffix.starts_with('-') || suffix.starts_with('_') || suffix.starts_with('.') {
                return true;
            }

            // Otherwise, check if the suffix is preceded by a separator
            let match_index = lower_key.len() - suffix.len();
            if match_index > 0 {
                let bytes = lower_key.as_bytes();
                let char_before = bytes[match_index - 1] as char;
                if char_before == '-' || char_before == '_' || char_before == '.' {
                    return true;
                }
            }
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_command() {
        let cases = vec![
            (
                vec!["curl", "-u", "user:password", "http://example.com"],
                "curl -u user:password http://example.com", // -u is not in sensitive list
            ),
            (
                vec!["myapp", "--password", "secret123"],
                "myapp --password [REDACTED]",
            ),
            (
                vec!["myapp", "--token=secret123"],
                "myapp --token=[REDACTED]",
            ),
            (
                vec!["myapp", "--api-key", "abcdef"],
                "myapp --api-key [REDACTED]",
            ),
            (
                vec!["db", "--db_password=secure"],
                "db --db_password=[REDACTED]",
            ),
            (
                vec!["service", "--aws_secret_access_key=XYZ"],
                "service --aws_secret_access_key=[REDACTED]", // matches _key suffix
            ),
            (
                vec!["ssh", "-p", "2222"],
                "ssh -p 2222", // -p is ambiguous, false in map
            ),
        ];

        for (input, expected) in cases {
            let input_vec: Vec<String> = input.iter().map(|s| s.to_string()).collect();
            assert_eq!(
                sanitize_command(&input_vec),
                expected,
                "Failed on input: {:?}",
                input
            );
        }
    }
}
